[UnifiedPDF] Handle PDF actions that move betwen pages and perform zooms.

sammygill opened the pull request:
#### f3ce152618e8b172daa501d4d579128f439dcfa7
<pre>
[UnifiedPDF] Handle PDF actions that move betwen pages and perform zooms.
<a href="https://bugs.webkit.org/show_bug.cgi?id=268508">https://bugs.webkit.org/show_bug.cgi?id=268508</a>
<a href="https://rdar.apple.com/122049363">rdar://122049363</a>

Reviewed by Tim Horton.

To start handling PDFActions we will check to see if an annotation has
an associated action. If it does then we check to see what type of
action it is. In this patch we provide support for actions that go to
next/previous pages, the first/last pages and zooming in/out.

We already have logic for these things so this is just a matter of
hooking up the PDFActions to them. For zooming in/out we hook into their
respective functions. For any sort page navigations we can use
scrollToPage. PDFDestination navigations are handled in a similar way.

Finally, I added in some logic to take into consideration page margins
in scrollToPointInPage.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::scrollToPointInPage):
(WebKit::UnifiedPDFPlugin::handlePDFActionForAnnotation):

Canonical link: <a href="https://commits.webkit.org/274598@main">https://commits.webkit.org/274598@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e76f792a2b58da5940cbe6319a31a41f560d0ff5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39538 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18517 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41892 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42072 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21399 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15846 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40112 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15603 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34227 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/13545 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43350 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35896 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35509 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39307 "Build is in progress. Recent messages:") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14350 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11824 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37567 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15956 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8851 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/16004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1391,7 +1394,10 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
     auto pointInContentsSpace = convertFromPageToContents(pointInPDFPageSpace, pageIndex);
-    scrollToPositionWithoutAnimation(pointInContentsSpace);
+    auto scaleFactor = this->scaleFactor() * documentFittingScale();
+    auto xMarginOffset = m_documentLayout.documentMargin.width()  * scaleFactor;
```
Can you replace all of this with something like 

`pointsInContentsSpace - m_documentLayout.documentMargin.scaled(scaleFactor)`?

aprotyas replied:
^ +1

aprotyas reviewed (commented):
LGTM

sammygill committed:
[UnifiedPDF] Handle PDF actions that move betwen pages and perform zooms.
https://bugs.webkit.org/show_bug.cgi?id=268508
rdar://122049363

Reviewed by Tim Horton.

To start handling PDFActions we will check to see if an annotation has
an associated action. If it does then we check to see what type of
action it is. In this patch we provide support for actions that go to
next/previous pages, the first/last pages and zooming in/out.

We already have logic for these things so this is just a matter of
hooking up the PDFActions to them. For zooming in/out we hook into their
respective functions. For any sort page navigations we can use
scrollToPage. PDFDestination navigations are handled in a similar way.

Finally, I added in some logic to take into consideration page margins
in scrollToPointInPage.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::scrollToPointInPage):
(WebKit::UnifiedPDFPlugin::handlePDFActionForAnnotation):

Canonical link: https://commits.webkit.org/274598@main

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -102,7 +102,7 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
 #if PLATFORM(MAC)
     RetainPtr<PDFAnnotation> nextTextAnnotation(AnnotationSearchDirection) const;
-    void handlePDFActionForAnnotation(PDFAnnotation *);
+    void handlePDFActionForAnnotation(PDFAnnotation *, unsigned currentPageIndex);
```
`PDFDocumentLayout::PageIndex` ?

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1499,7 +1499,7 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
 {
     auto pointInContentsSpace = convertFromPageToContents(pointInPDFPageSpace, pageIndex);
-    scrollToPositionWithoutAnimation(pointInContentsSpace);
+    scrollToPositionWithoutAnimation(pointInContentsSpace - m_documentLayout.documentMargin.scaled(this->scaleFactor() * documentFittingScale()));
```
Ideally we'd add a helper for whatever coordinate transform this is.
